proposals/0000-scan-takewhile-dropwhile.md

```diff
@@ -0,0 +1,222 @@
+  /// `true` and returning the remainder.
+  func dropWhile(@noescape dropElement: (Self.Generator.Element) throws -> Bool) rethrows -> Self.SubSequence
+  /// Returns a subsequence containing the elements until `takeElement` returns
+  /// `false` and skipping the remainder.
+  func takeWhile(@noescape takeElement: (Self.Generator.Element) throws -> Bool) rethrows -> Self.SubSequence
+}
+```
+
+Also provide default implementations on `SequenceType` that return
+`AnySequence`, and default implementations on `CollectionType` that return a
+slice.
+
+`LazySequenceType` and `LazyCollectionType` will also be extended with
+implementations of `scan(_:combine:)`, `dropWhile(_:)`, and `takeWhile(_:)`
+that return lazy sequence/collection types. Like the lazy `filter(_:)`,
+`dropWhile(_:)` will perform the filtering when `startIndex` is accessed.
+
+Add a global function:
+
+```swift
+/// Returns an infinite sequence of lazy applications of `apply` to the
+/// previous value. For example:
+///
+///     iterate(1, apply: { $0 * 2 }) // yields: 1, 2, 4, 8, 16, 32, 64, ...
+func iterate<T>(initial: T, apply: T -> T) -> IterateSequence<T>
+```
+
+## Detailed design
+
+In addition to the above declarations, provide default implementations based on
+`AnySequence`, similarly to how functions like `dropFirst(_:)` and `prefix(_:)`
+are handled:
+
+```swift
+extension SequenceType {
+  func dropWhile(@noescape dropElement: (Self.Generator.Element) throws -> Bool) rethrows -> AnySequence<Self.Generator.Element>
+  func takeWhile(@noescape takeElement: (Self.Generator.Element) throws -> Bool) rethrows -> AnySequence<Self.Generator.Element>
+}
+```
+
+These default implementations produce an `AnySequence` that wraps an `Array`
+(as the functions must be implemented eagerly to match expected behavior).
+
+Provide default implementations on `CollectionType` as well:
+
+```swift
+extension CollectionType {
+  func dropWhile(@noescape dropElement: (Self.Generator.Element) throws -> Bool) rethrows -> Self.SubSequence
+  func takeWhile(@noescape takeElement: (Self.Generator.Element) throws -> Bool) rethrows -> Self.SubSequence
+}
+```
+
+Extend `LazySequenceType` with lazy versions of the functions:
+
+```swift
+extension LazySequenceType {
+  func scan<T>(initial: T, combine: (T, Self.Generator.Element) -> T) -> LazyScanSequence<Self.Elements, T>
+  func dropWhile(dropElement: (Self.Generator.Element) -> Bool) -> LazyDropWhileSequence<Self.Elements>
+  func takeWhile(takeElement: (Self.Generator.Element) -> Bool) -> LazyTakeWhileSequence<Self.Elements>
+}
+```
+
+The types `LazyScanSequence`, `LazyDropWhileSequence`, and
+`LazyTakeWhileSequence` all conform to `LazySequenceType`.
+
+Extend `LazyCollectionType` with collection variants for `dropWhile(_:)` and
+`takeWhile(_:)` (but not `scan(_:combine:)` because there's no way to recover
+the value from an index without re-scanning from the start):
```

Comment:
That also requires that the index have a reference to the underlying collection as well, so it can increment itself, which requires that the collection itself be held by-reference by the collection wrapper (it could be stored inline and just copied into each index, but arbitrary collections might be quite large so that would be an expensive index).

I seem to recall seeing at one point a proposal to change how collection indexes work so the get incremented/decremented via methods on their associated collection. I don't see it in the swift-evolution repo so I assume it was somewhere on the ML. In any case, given a change like that, it would be a lot easier to store the value in the index because it wouldn't have to keep track of the associated underlying collection.